source/azure_iot_hub_client.c

```diff
@@ -460,6 +461,72 @@ static AzureIoTHubClientResult_t prvWaitForSubAck( AzureIoTHubClient_t * pxAzure
 }
 /*-----------------------------------------------------------*/
 
+/**
+ * Generate the SAS token based on :
+ *   https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security#use-a-shared-access-policy
+ **/
+static uint32_t prvIoTHubClientGetToken( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                         uint64_t ullExpiryTimeSecs,
+                                         const uint8_t * ucKey,
+                                         uint32_t ulKeyLen,
+                                         uint8_t * pucSASBuffer,
+                                         uint32_t ulSasBufferLen,
+                                         uint32_t * pulSaSLength )
+{
+    uint8_t * pucHMACBuffer;
+    uint32_t ulHMACBufferLength = azureiothubHMACBufferLength;
+    az_span xSpan = az_span_create( pucSASBuffer, ( int32_t ) ulSasBufferLen );
+    az_result xCoreResult;
+    uint32_t ulBytesUsed;
+    uint32_t ulBufferLeft;
+    size_t xLength;
+
+    xCoreResult = az_iot_hub_client_sas_get_signature( &( pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore ),
+                                                       ullExpiryTimeSecs, xSpan, &xSpan );
+
+    if( az_result_failed( xCoreResult ) )
+    {
+        AZLogError( ( "AzureIoTHubClient failed to get signature with error status: %08x", xCoreResult ) );
+        return eAzureIoTHubClientFailed;
+    }
+
+    ulBytesUsed = ( uint32_t ) az_span_size( xSpan );
+    ulBufferLeft = ulSasBufferLen - ulBytesUsed;
+
+    if( ulBufferLeft < ulHMACBufferLength )
+    {
+        AZLogError( ( "AzureIoTHubClient token generation failed with insufficient buffer size." ) );
+        return eAzureIoTHubClientOutOfMemory;
+    }
+
+    ulBufferLeft -= ulHMACBufferLength;
+    pucHMACBuffer = pucSASBuffer + ulSasBufferLen - ulHMACBufferLength;
```

Comment:
I think a comment here would help.  Basically you're using the last 48 bytes of the app's passed buffer to store the HMAC, the first part of the buffer is the result of the `az_iot_hub_client_sas_get_signature` call, and there's a middle section of buffer between end of `az_iot_hub_client_sas_get_signature` output and before the reserved 48 bytes that you need for AzureIoT_Base64HMACCalculate`s `pucBuffer` parameter, correct?

This all looks good, but it took me a bit to figure out how it all fits together.  So a comment about "only want caller to have to pass one buffer so we use different portions of the space as follows..." by time we're at HMAC time would've made this easier.